### PR TITLE
KAFKA-12895: Drop support for Scala 2.12 in Kafka 4.0

### PR DESCRIPTION
**Pull request by frankvicky**
JIRA: [KAFKA-12895](https://issues.apache.org/jira/browse/KAFKA-12895)

> We propose to deprecate Scala 2.12 support n Apache Kafka 3.0 and to drop it in Apache Kafka 4.0.

Implement [KIP-751](https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=181308218)

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by frankvicky** [comment]
Hi @jolshan,
You mentioned earlier that you were interested in this ticket. Could you please take a look? 
Thank you ! 😺

**Comment by m1a2st** [comment]
Thanks for your PR, left some comments, PTAL

**Comment by m1a2st** [comment]
Please also update the LICENSE-binary, there are some scala 2.12.19 version, and should also update the gradlewAll file.
Some html document also need to update.

**Comment by frankvicky** [comment]
Currently there are some html metiond about Scala 2.12 are located in `Streams` module. I'm not sure they should be change in this PR or not.

File: gradle/dependencies.gradle [hunk 1]
```diff
@@ -22,27 +22,24 @@ ext {
   libs = [:]
 
   // Available if -PscalaVersion is used. This is useful when we want to support a Scala version that has
-  // a higher minimum Java requirement than Kafka. This was previously the case for Scala 2.12 and Java 7.
-  availableScalaVersions = [ '2.12', '2.13' ]
+  // a higher minimum Java requirement than Kafka.
+  availableScalaVersions = [ '2.13' ]
```
Comment (chia7712) [inline gradle/dependencies.gradle]: Is it still used?

**Review by chia7712 — COMMENTED** [review]
@frankvicky Could you please add this change to `upgrade.html`?

File: core/src/main/scala/kafka/coordinator/transaction/TransactionMarkerChannelManager.scala [hunk 2]
```diff
@@ -121,7 +120,7 @@ class TxnMarkerQueue(@volatile var destination: Node) extends Logging {
   }
 
   def addMarkers(txnTopicPartition: Int, pendingCompleteTxnAndMarker: PendingCompleteTxnAndMarkerEntry): Unit = {
-    val queue = CoreUtils.atomicGetOrUpdate(markersPerTxnTopicPartition, txnTopicPartition, {
+    val queue =markersPerTxnTopicPartition.getOrElseUpdate(txnTopicPartition, {
```
Comment (mumrah) [inline core/src/main/scala/kafka/coordinator/transaction/TransactionMarkerChannelManager.scala]: nit: whitespace

**Review by mumrah — COMMENTED** [review]
For the classes where we're actually change code (LogCleanerManager and LogCleaner), can you double check that we have unit test code coverage for these methods? I just want to be sure we're not changing the logic.

**Comment by jolshan** [comment]
Scanned through some of the files here: https://github.com/search?q=repo%3Aapache%2Fkafka%202.12&type=code

I noticed the vagrant file also has 2.12 references. Is this just for building old versions? https://github.com/apache/kafka/blob/9cb49092f0865ae18487444be52aa1d03b915de3/vagrant/base.sh#L116

What is the concern about the streams changes?

**Comment by chia7712** [comment]
> I noticed the vagrant file also has 2.12 references. Is this just for building old versions? 

yes, the dockerfile (e2e) also has 2.12 references. 

https://github.com/apache/kafka/blob/6ee46ea86530e019798ba0d4183164d7925dde63/tests/docker/Dockerfile#L99

File: core/src/test/scala/integration/kafka/api/BaseConsumerTest.scala [hunk 3]
```diff
@@ -117,38 +116,29 @@ object BaseConsumerTest {
...
+      Arguments.of("kraft", "classic"),
+      Arguments.of("kraft", "consumer")
+    )
   }
 
-  // In Scala 2.12, it is necessary to disambiguate the java.util.stream.Stream.of() method call
-  // in the case where there's only a single Arguments in the list. The following commented-out
-  // method works in Scala 2.13, but not 2.12. For this reason, tests which run against just a
-  // single combination are written using @CsvSource rather than the more elegant @MethodSource. 
-  // def getTestQuorumAndGroupProtocolParametersZkOnly() : java.util.stream.Stream[Arguments] = {
```
Comment (chia7712) [inline core/src/test/scala/integration/kafka/api/BaseConsumerTest.scala:131]: This should be used by `testAssignAndConsumeWithLeaderChangeValidatingPositions`. Since we are phasing out ZooKeeper, it's fine to remove it for now.

https://github.com/apache/kafka/blob/6ee46ea86530e019798ba0d4183164d7925dde63/core/src/test/scala/integration/kafka/api/PlaintextConsumerAssignTest.scala#L143

**Comment by chia7712** [comment]
> For the classes where we're actually change code (LogCleanerManager and LogCleaner), can you double check that we have unit test code coverage for these methods? I just want to be sure we're not changing the logic.

This PR should primarily focus on removing Scala 2.12. The changes made to `LogCleanerManager` and `LogCleaner` seem more like cosmetic modifications. It might be better to revert them to their original state.

File: core/src/main/scala/kafka/coordinator/transaction/TransactionMarkerChannelManager.scala (hunk 2 above)
Comment (frankvicky) [reply core/src/main/scala/kafka/coordinator/transaction/TransactionMarkerChannelManager.scala]: Thank you, fixed it.

Comment (frankvicky) [reply core/src/main/scala/kafka/coordinator/transaction/TransactionMarkerChannelManager.scala]: > For the classes where we're actually change code (LogCleanerManager and LogCleaner), can you double check that we have unit test code coverage for these methods? I just want to be sure we're not changing the logic.

Regarding to comment from @chia7712 , I have reverted the changes in `LogCleanerManager` and `LogCleaner`

File: core/src/test/scala/integration/kafka/api/BaseConsumerTest.scala (hunk 3 above)
Comment (frankvicky) [reply core/src/test/scala/integration/kafka/api/BaseConsumerTest.scala:131]: I have uncommented the `getTestQuorumAndGroupProtocolParametersZkOnly` method and replaced `@CsvSource` with `@MethodSource`.

File: gradle/dependencies.gradle (hunk 1 above)
Comment (frankvicky) [reply gradle/dependencies.gradle]: Nope, I will remove it

File: gradlewAll [hunk 4]
```diff
@@ -17,5 +17,5 @@
 
 # Convenient way to invoke a gradle command with all Scala versions supported
 # by default
```
Comment (ijuma) [inline gradlewAll:19]: We should probably add a comment here that this is no longer needed, but we are keeping it for backwards compatibility.

File: README.md [hunk 5]
```diff
@@ -11,8 +11,7 @@ the broker and tools has been deprecated since Apache Kafka 3.7 and removal of b
 see [KIP-750](https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=181308223) and
 [KIP-1013](https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=284789510) for more details).
 
-Scala 2.12 and 2.13 are supported and 2.13 is used by default. Scala 2.12 support has been deprecated since
-Apache Kafka 3.0 and will be removed in Apache Kafka 4.0 (see [KIP-751](https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=181308218)
+Scala 2.13 is used by default. Scala 2.12 support has been removed in Apache Kafka 4.0 (see [KIP-751](https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=181308218)
```
Comment (ijuma) [inline README.md]: I think we should simply say that Scala 2.13 is the only version we support. The readme is not a changelog. We have a separate file for the changelog.

File: README.md [hunk 6]
```diff
@@ -114,14 +113,14 @@ Using compiled files:
 ### Cleaning the build ###
     ./gradlew clean
 
-### Running a task with one of the Scala versions available (2.12.x or 2.13.x) ###
+### Running a task with Scala 2.13.x ###
```
Comment (ijuma) [inline README.md]: I think we should just delete this section and all the other sections related to multiple Scala versions, etc.

File: core/src/main/scala/kafka/utils/CoreUtils.scala [hunk 7]
```diff
@@ -242,31 +241,6 @@ object CoreUtils {
...
-  /**
-   * Atomic `getOrElseUpdate` for concurrent maps. This is optimized for the case where
-   * keys often exist in the map, avoiding the need to create a new value. `createValue`
-   * may be invoked more than once if multiple threads attempt to insert a key at the same
-   * time, but the same inserted value will be returned to all threads.
-   *
-   * In Scala 2.12, `ConcurrentMap.getOrElse` has the same behaviour as this method, but JConcurrentMapWrapper that
-   * wraps Java maps does not.
-   */
-  def atomicGetOrUpdate[K, V](map: concurrent.Map[K, V], key: K, createValue: => V): V = {
```
Comment (ijuma) [inline core/src/main/scala/kafka/utils/CoreUtils.scala]: Can we double check that this comment is accurate for the various `concurrent.Map` implementations?

**Review by ijuma — COMMENTED** [review]
Thanks for the PR, I left a few comments.

**Comment by ijuma** [comment]
> This PR should primarily focus on removing Scala 2.12. The changes made to LogCleanerManager and LogCleaner seem more like cosmetic modifications. It might be better to revert them to their original state.

These are not just cosmetic, they're fixing unnecessary perf overhead as well. I don't think we should have _any_ code remaining that has a "Scala 2.12" comment. So, the change was fine, in my opinion. If we want to do it via a separate PR, let's make sure we have a JIRA and we don't consider the work done until this is also completed.

File: gradlewAll (hunk 4 above)
Comment (chia7712) [reply gradlewAll:19]: Maybe we should print deprecated information to users to remind that this `gradlewAll` is unnecessary and will be removed later.

**Comment by chia7712** [comment]
> These are not just cosmetic, they're fixing unnecessary perf overhead as well. I don't think we should have any code remaining that has a "Scala 2.12" comment. So, the change was fine, in my opinion. If we want to do it via a separate PR, let's make sure we have a JIRA and we don't consider the work done until this is also completed.

It seems to me that both "cosmetic" changes and "unnecessary performance overhead" should be handled in a follow-up to avoid prolonged discussions. I believe there are several improvements, such as KAFKA-14615, that can be addressed after dropping Scala 2.12.

**Comment by chia7712** [comment]
> If we want to do it via a separate PR, let's make sure we have a JIRA and we don't consider the work done until this is also completed.

https://issues.apache.org/jira/browse/KAFKA-17667

**Comment by ijuma** [comment]
My point is a bit different. It's one thing to have optimizations, but it's another to have leftover code like:

> // Note: we don't use retain or filterInPlace method in this function because retain is deprecated in
> // scala 2.13 while filterInPlace is not available in scala 2.12.

Doing what it takes to clean up such comments should be part of removing 2.12 support.

File: README.md (hunk 6 above)
Comment (frankvicky) [reply README.md:116]: I've kept the section related to ./gradlewAll since we haven't removed it yet.

File: README.md (hunk 5 above)
Comment (frankvicky) [reply README.md]: I’ve kept the section related to `./gradlewAll` since we haven’t removed it yet.

File: core/src/main/scala/kafka/utils/CoreUtils.scala (hunk 7 above)
Comment (frankvicky) [reply core/src/main/scala/kafka/utils/CoreUtils.scala]: I have reviewed the implementation in the Scala source code.

We could remove `atomicGetOrUpdate()` in Scala 2.13, as Scala now provides a very similar implementation for `getOrElseUpdate()`.

In Scala 2.12, if you convert a ConcurrentHashMap using `scala.collection.JavaConverters.asScala`, you will indeed get a less reliable version of `getOrElseUpdate` that does not handle race conditions or atomic operations properly. This is the less robust version, as it doesn't ensure atomic updates.

2.13
```scala
  /**
    * Replaces the entry for the given key only if it was previously mapped
    * to some value.
    *
    * $atomicop
    *
    * @param k   key for which the entry should be replaced
    * @param v   value to be associated with the specified key
    * @return    `Some(v)` if the given key was previously mapped to some value `v`, or `None` otherwise
    */
  def replace(k: K, v: V): Option[V]

  override def getOrElseUpdate(key: K, @deprecatedName("op", since="2.13.13") defaultValue: => V): V = get(key) match {
    case Some(v) => v
    case None =>
      val v = defaultValue
      putIfAbsent(key, v) match {
        case Some(ov) => ov
        case None => v
      }
  }
```

2.12
```scala
  /** If given key is already in this map, returns associated value.
   *
   *  Otherwise, computes value from given expression `op`, stores with key
   *  in map and returns that value.
   *
   *  Concurrent map implementations may evaluate the expression `op`
   *  multiple times, or may evaluate `op` without inserting the result.
   *  
   *  @param  key the key to test
   *  @param  op  the computation yielding the value to associate with `key`, if
   *              `key` is previously unbound.
   *  @return     the value associated with key (either previously or as a result
   *              of executing the method).
   */
  def getOrElseUpdate(key: K, op: => V): V =
    get(key) match {
      case Some(v) => v
      case None => val d = op; this(key) = d; d
    }

```

Comment (chia7712) [reply core/src/main/scala/kafka/utils/CoreUtils.scala]: @frankvicky we should check the `JavaCollectionWrappers` for 2.13

https://github.com/scala/scala/blob/2.13.x/src/library/scala/collection/convert/JavaCollectionWrappers.scala#L482

```scala
    override def getOrElseUpdate(key: K, op: => V): V =
      underlying.computeIfAbsent(key, _ => op) match {
        case null => super/*[concurrent.Map]*/.getOrElseUpdate(key, op)
        case v    => v
      }
```

Hence, the wrapper actually leverage `computeIfAbsent` for normal case. The `getOrElseUpdate` is used to put the `null` to map since `computeIfAbsent` does not. That is for following `getOrElseUpdate` semantics

The `getOrElseUpdate` of current scala 2.13 is good, and we can do a bit cleanup (in the follow-up) as current usages (`markersQueuePerBroker` and `markersPerTxnTopicPartition`) are not necessary to use scala wrapper ..

**Comment by chia7712** [comment]
> Doing what it takes to clean up such comments should be part of removing 2.12 support.

I agree with cleaning up the comments and straightforwardly rewriting the code using Scala 2.13 APIs. However, any tasks requiring further discussion or additional testing can be deferred to a follow-up. This approach helps prevent infinite conflicts in this large patch (with 54 files changed)

File: README.md (hunk 6 above)
Comment (frankvicky) [reply README.md]: I’ve kept the section related to ./gradlewAll since we haven’t removed it yet.

File: core/src/main/scala/kafka/utils/CoreUtils.scala (hunk 7 above)
Comment (ijuma) [reply core/src/main/scala/kafka/utils/CoreUtils.scala]: The Java implementation of `computeIfAbsent` acquires locks, so the behavior is different.

File: README.md (hunk 6 above)
Comment (ijuma) [reply README.md]: I would remove that too. We can simply add a comment to `gradlewAll` explaining that it has been kept for compatibility reasons, but there is no reason to use it anymore.

File: core/src/main/scala/kafka/utils/CoreUtils.scala (hunk 7 above)
Comment (chia7712) [reply core/src/main/scala/kafka/utils/CoreUtils.scala]: > The Java implementation of computeIfAbsent acquires locks, so the behavior is different.

Pardon me, either `computeIfAbsent` or `putIfAbsent` acquires lock (`ConcurrentHashMap` impl) when it needs to mutable the inner table. Additionally, the main behavior difference is that `atomicGetOrUpdate` has a side effect that it may create unnecessary value in race condition, but `getOrElseUpdate`  (`computeIfAbsent`) does not. For another, the main concern in this use case is the optimized read (according to comment). Maybe we can run JMH to check the perf of read-heavy case between `computeIfAbsent` and `get`. 

@frankvicky Could you please add a test case to `ConcurrentMapBenchmark`?

Comment (frankvicky) [reply core/src/main/scala/kafka/utils/CoreUtils.scala]: Test in macOSX
```
Benchmark                                                            (mapSize)  (writePercentage)  Mode  Cnt  Score   Error  Units
ConcurrentMapBenchmark.testConcurrentHashMapComputeIfAbsentReadOnly        100                0.1  avgt    5  4.706 ± 0.074  ns/op
ConcurrentMapBenchmark.testConcurrentHashMapGetReadOnly                    100                0.1  avgt    5  3.557 ± 0.032  ns/op
JMH benchmarks done
```

Comment (frankvicky) [reply core/src/main/scala/kafka/utils/CoreUtils.scala]: Test in Ubuntu
```
Benchmark                                                            (mapSize)  (writePercentage)  Mode  Cnt  Score   Error  Units
ConcurrentMapBenchmark.testConcurrentHashMapComputeIfAbsentReadOnly        100                0.1  avgt    5  4.099 ± 0.080  ns/op
ConcurrentMapBenchmark.testConcurrentHashMapGetReadOnly                    100                0.1  avgt    5  3.439 ± 0.084  ns/op
JMH benchmarks done
```

Comment (ijuma) [reply core/src/main/scala/kafka/utils/CoreUtils.scala]: @chia7712 Right, one implementation _always_ acquires a lock to ensure the value is only computed once while the other only acquires locks if it needs to mutate the underlying map. The latter is better for read-heavy while the former is better for write-heavy cases. I think it would be best to have changes that may make things worse submitted separately. This PR should only have the changes that are neutral or always better.

Comment (chia7712) [reply core/src/main/scala/kafka/utils/CoreUtils.scala]: >  I think it would be best to have changes that may make things worse submitted separately. This PR should only have the changes that are neutral or always better.

totally agree. open https://issues.apache.org/jira/browse/KAFKA-17677 as follow-up. @frankvicky could you please keep `atomicGetOrUpdate`?

**Comment by frankvicky** [comment]
I accidentally deleted the branch due to an operational mistake, but I’ve restored it now. 
Apologies for the confusion. 🙇🏼

File: docs/upgrade.html [hunk 8]
```diff
@@ -26,6 +26,10 @@ <h5><a id="upgrade_400_notable" href="#upgrade_400_notable">Notable changes in 4
             <ul>
                 <li><b>Common</b>
                     <ul>
+                        <li>
```
Comment (chia7712) [inline docs/upgrade.html]: Could you please move this to the 'Other Changes' section? This section seems to be focused on configuration changes.

File: core/src/main/scala/kafka/utils/json/DecodeJson.scala [hunk 9]
```diff
@@ -17,10 +17,8 @@
 
 package kafka.utils.json
 
-import scala.collection.{Map, Seq}
-import scala.collection.compat._
+import scala.collection.{Factory, Map, Seq}
```
Comment (chia7712) [inline core/src/main/scala/kafka/utils/json/DecodeJson.scala:20]: please revert this unrelated changes

**Review by chia7712 — COMMENTED** [review]
@frankvicky thanks for this patch.

Please update the command from `gradlewAll` to `gradlew`

https://github.com/apache/kafka/blob/trunk/release/release.py#L329
https://github.com/apache/kafka/blob/trunk/release/release.py#L354

File: core/src/main/scala/kafka/utils/json/DecodeJson.scala (hunk 9 above)
Comment (frankvicky) [reply core/src/main/scala/kafka/utils/json/DecodeJson.scala:20]: We are getting rid of `scala.collection.compat._` since it was a workaround package for cross-building with Scala 2.12. Now, we can import Factory directly from `scala.collection`.

File: core/src/test/java/kafka/admin/ConfigCommandTest.java [hunk 10]
```diff
@@ -1013,7 +1013,7 @@ public void shouldAlterTopicConfig(boolean file) {
         addedConfigs.put("delete.retention.ms", "1000000");
         addedConfigs.put("min.insync.replicas", "2");
         if (file) {
-            File f = kafka.utils.TestUtils.tempPropertiesFile(JavaConverters.mapAsScalaMap(addedConfigs));
+            File f = kafka.utils.TestUtils.tempPropertiesFile(CollectionConverters.MapHasAsScala(addedConfigs).asScala());
```
Comment (chia7712) [inline core/src/test/java/kafka/admin/ConfigCommandTest.java]: Could you please import `scala.jdk.javaapi.CollectionConverters` instead then rewrite this to `CollectionConverters.asScala(addedConfigs)`

File: core/src/test/java/kafka/admin/ConfigCommandTest.java [hunk 11]
```diff
@@ -2294,6 +2294,6 @@ public AlterClientQuotasResult alterClientQuotas(Collection<ClientQuotaAlteratio
 
     @SuppressWarnings({"deprecation"})
     private <T> Seq<T> seq(Collection<T> seq) {
-        return JavaConverters.asScalaIteratorConverter(seq.iterator()).asScala().toSeq();
+        return IteratorHasAsScala(seq.iterator()).asScala().toSeq();
```
Comment (chia7712) [inline core/src/test/java/kafka/admin/ConfigCommandTest.java]: `return CollectionConverters.asScala(seq).toSeq();`

File: core/src/test/java/kafka/admin/ConfigCommandTest.java [hunk 12]
```diff
@@ -2294,6 +2294,6 @@ public AlterClientQuotasResult alterClientQuotas(Collection<ClientQuotaAlteratio
 
     @SuppressWarnings({"deprecation"})
```
Comment (chia7712) [inline core/src/test/java/kafka/admin/ConfigCommandTest.java]: please remove unnecessary annotation

**Review by chia7712 — APPROVED** [review]
LGTM

I'm going to merge this PR. Please feel free to leave any comments and follow [KAFKA-17667](https://issues.apache.org/jira/browse/KAFKA-17667) for further optimizations and cleanup.
